### PR TITLE
Commutative rules for DNN subgraphs fusion

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

related: https://github.com/opencv/opencv/pull/24463#issuecomment-1783033931

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by fengyuentau — COMMENTED** [review]
Overall looks good to me. I plan to work on a test for subgraph fusion this weekend (load a net and see whether it can be simplified to the expected dnn layer). Could you wait a bit for that?

**Comment by fengyuentau** [comment]
Also note this warning:

```
/build/precommit_linux64/4.x/opencv/modules/dnn/src/graph_simplifier.hpp:48:25: warning: inline function 'virtual bool cv::dnn::ImportGraphWrapper::isCommutativeOp(const string&) const' used but never defined
```

**Comment by dkurt** [comment]
@fengyuentau , sure, thanks!

**Comment by fengyuentau** [comment]
> a test for subgraph fusion

While I am making this, I found there are two patterns for softplus, which can be merged into one in this PR:

https://github.com/opencv/opencv/blob/ed52f7feeaae6f0e96ee4dc36ae5f9fb4a6d959c/modules/dnn/src/onnx/onnx_graph_simplifier.cpp#L988-L1014

The only difference of these two subgraphs is

```
         int add = addNodeToMatch("Add", addVal, exp); 

         int add = addNodeToMatch("Add", exp, addVal); 
```

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -2780,4 +2780,11 @@ TEST_P(Test_ONNX_layers, Expand_shape_model4) {
 
 INSTANTIATE_TEST_CASE_P(/**/, Test_ONNX_nets, dnnBackendsAndTargets());
 
+TEST(Test_ONNX_importer, commutative_fusion) {
+    Net net = readNet(_tf("models/mish.onnx"));
+    std::vector<std::string> layerTypes;
+    net.getLayerTypes(layerTypes);
+    ASSERT_TRUE(std::find(layerTypes.begin(), layerTypes.end(), "Mish") != layerTypes.end());
+}
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2788]: Move after https://github.com/opencv/opencv/pull/24500

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 2]
```diff
@@ -125,15 +125,41 @@ class ONNXGraphWrapper : public ImportGraphWrapper
...
+    {
+        return type == "Add" || type == "Mul" || type == "Equal" || type == "Max";
     }
 
 private:
     int numInputs, numInitializers;
     opencv_onnx::GraphProto& net;
 };
 
+static float extractConstant(const Ptr<ImportGraphWrapper>& net, int node_id, int input_id)
```
Comment (fengyuentau) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:142]: I propose to template this function. We also need to extract int.

Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp]: Ok, it also works.

**Comment by fengyuentau** [comment]
These warnings from `default Win64` are not related to changes in this PR:

```
C:\build\precommit_windows64\4.x\opencv\modules\videoio\test\test_gstreamer.cpp(193): warning C4244: 'initializing': conversion from 'int' to 'float', possible loss of data [C:\build\precommit_windows64\build\modules\videoio\opencv_test_videoio.vcxproj]
C:\build\precommit_windows64\4.x\opencv\modules\videoio\test\test_gstreamer.cpp(193): warning C4244: 'initializing': conversion from 'int' to 'const float', possible loss of data [C:\build\precommit_windows64\build\modules\videoio\opencv_test_videoio.vcxproj]
```

**Review by fengyuentau — APPROVED** [review]

**Pull request merged**